discovery_engine_core/web-api/src/models.rs

```diff
@@ -94,14 +77,17 @@ impl From<DocumentId> for String {
     }
 }
 
+fn is_valid_id(id: &str) -> bool {
+    static RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[a-zA-Z0-9_\-:@.]+$").unwrap());
+    RE.is_match(id)
+}
+
 impl DocumentId {
     pub fn new(id: impl Into<String>) -> Result<Self, Error> {
         let id = id.into();
 
-        if id.is_empty() {
-            Err(Error::DocumentIdEmpty)
-        } else if id.contains('\u{0000}') {
-            Err(Error::DocumentIdContainsNul)
+        if !is_valid_id(&id) {
+            Err(Error::InvalidDocumentId { invalid_id: id })
         } else {
             Ok(Self(id))
         }
```

Comment:
:art: since only one if-branch remains, we could simplify the doubled negation. same for the other id constructors.
```suggestion
        if is_valid_id(&id) {
            Ok(Self(id))
        } else {
            Err(Error::InvalidDocumentId { invalid_id: id })
        }
```